educe/ptb/head_finder.py

```diff
@@ -0,0 +1,150 @@
+"""This submodule provides several functions that find heads in trees.
+
+It uses head rules as described in (Collins 1999), Appendix A.
+See `http://www.cs.columbia.edu/~mcollins/papers/heads`,
+Bikel's 2004 CL paper on the intricacies of Collins' parser
+and the classes in (StanfordNLP) CoreNLP that inherit from
+`AbstractCollinsHeadFinder.java` .
+"""
+
+from __future__ import print_function
+
+
+import os
+
+from nltk import Tree
+
+
+HEAD_RULES_FILE = os.path.join(os.path.dirname(__file__),
+                               'collins_head_rules')
+
+
+def load_head_rules(f):
+    """Load the head rules from file f.
+
+    Return a dictionary from parent non-terminal to (direction,
+    priority list).
+    """
+    rules = dict()
+    with open(f) as f:
+        for line in f:
+            prnt_nt, drctn, prrty_lst = line.strip().split('\t')
+            lbls = prrty_lst.split() if prrty_lst != '_' else []
+            rules[prnt_nt] = (drctn, lbls)
+    return rules
+
+
+HEAD_RULES = load_head_rules(HEAD_RULES_FILE)
+
+
+# helper functions
+def _find_head_generic(direction, priority_list, cnt_hws):
+    """Determine the head word of a phrase"""
+    if direction == 'Left':  # scan children left to right
+        cands = list(enumerate(cnt_hws))
+    elif direction == 'Right':
+        cands = list(reversed(list(enumerate(cnt_hws))))
+    else:
+        err_msg = 'Direction can obly be Left or Right, got {}'
+        raise ValueError(err_msg.format(direction))
+
+    # try to find each label in the priority list, in turn
+    # if none were found, return the head word of the left-
+    # (resp. right-) most child
+    for lbl in priority_list:
+        for c_idx, (cnt, hw) in cands:
+            if cnt == lbl:
+                return (c_idx, hw)
+    else:
+        c_idx, (cnt, hw) = cands[0]
+        return (c_idx, hw)
+
+
+def _find_head_np(cnt_hws):
+    """Find head word in NP following specific rules"""
+    cands = list(enumerate(cnt_hws))
```

Comment:
If I understand correctly you have ranked sets of categories, 

* some of which you want to search rightwards for,
* some leftwards
* some only the last word

You want to return the head word from the best ranked set.  Naturally my first instinct is to want to separate out this logic from the ranks, but I can see that this gets a bit tricky with the different l/r logic.